packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,252 @@
+    }
+}
+
+// A number of MPI routines (in the current OpenMPI implementation at least) are
+// checking at runtime that all pointers passed to them as argument are not null
+// and call program termination even when no data would actually be
+// communicated. This is problematic when using std::vector to store buffers and
+// other arrays because std::vector::data() may return a null pointer if the
+// vector is empty. This function "swaps" the pointer with some other pointer
+// that is not dereferenceable if it is the null pointer.
+template <typename T>
+static T *notNullPtr( T *p )
+{
+    return p ? p : reinterpret_cast<T *>( -1 );
+}
+
+class Distributor
+{
+  public:
+    Distributor( MPI_Comm comm )
+        : _comm( comm )
+    {
+    }
+
+    template <typename View>
+    size_t createFromSends( View const &destination_ranks )
+    {
+        static_assert( View::rank == 1, "" );
+        static_assert(
+            std::is_same<typename View::non_const_value_type, int>::value, "" );
+        int comm_rank;
+        MPI_Comm_rank( _comm, &comm_rank );
+
+        _permute = Kokkos::View<int *, Kokkos::HostSpace>(
+            Kokkos::ViewAllocateWithoutInitializing( "permute" ),
+            destination_ranks.size() );
+        std::vector<int> destinations;
+        sortAndDetermineBufferLayout( destination_ranks, _permute, destinations,
+                                      _dest_counts, _dest_offsets );
+
+        {
+            int const reorder = 0; // Ranking may *not* be reordered
+            int const sources[1] = {comm_rank};
+            int const degrees[1] = {static_cast<int>( destinations.size() )};
+
+            MPI_Dist_graph_create(
+                _comm, 1, sources, degrees, notNullPtr( destinations.data() ),
+                MPI_UNWEIGHTED, MPI_INFO_NULL, reorder, &_comm_dist_graph );
+        }
+
+        int indegrees;
+        int outdegrees;
+        int weighted;
+        MPI_Dist_graph_neighbors_count( _comm_dist_graph, &indegrees,
+                                        &outdegrees, &weighted );
+        DTK_ENSURE( weighted == 0 );
+        DTK_ENSURE( outdegrees == static_cast<int>( destinations.size() ) );
+
+        std::vector<int> sources( indegrees );
+        MPI_Dist_graph_neighbors( _comm_dist_graph, indegrees,
+                                  notNullPtr( sources.data() ), MPI_UNWEIGHTED,
+                                  outdegrees, notNullPtr( destinations.data() ),
+                                  MPI_UNWEIGHTED );
+
+        _src_counts.resize( indegrees );
+        MPI_Neighbor_alltoall( _dest_counts.data(), 1, MPI_INT,
+                               _src_counts.data(), 1, MPI_INT,
+                               _comm_dist_graph );
+
+        _src_offsets.resize( indegrees + 1 );
+        // exclusive scan
+        _src_offsets[0] = 0;
+        for ( int i = 0; i < indegrees; ++i )
+            _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+
+#if defined( REORDER_RECV )
```

Comment:
So is this to reorder the output of the graph communication into a preferred order?

Reply:
Reordering the receive buffer.  Some of the downstream code (by that I mean packages other than `Search`) seems to be sensitive to it.  This needs more investigation.  My goal is to get rid of it.